<img src = "https://global-uploads.webflow.com/5f0d53c042a9ed6288de7f8d/5f6337ae2cfaa10946ceeb06_Hacktiv8%20logo%20horizontal%2001%20black-p-500.png" width = 400>
<h1 align=center><font size = 5>Hacktiv8 PTP Introduction to Data Science Projects 2 // Statistical Treatment for Datasets</font></h1>

Title: Hacktiv8 PTP Introduction to Data Science Projects 2: Statistical Treatment for Datasets Starter 

## NYC Property Sales Introduction

The aim of this projects is to introduce you to practical statistic with Python as concrete and as consistent as possible. Using what you’ve learned; download the NYC Property Sales Dataset from Kaggle. This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period.

This dataset contains the location, address, type, sale price, and sale date of building units sold. A reference on the trickier fields:

* `BOROUGH`: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).
* `BLOCK`; `LOT`: The combination of borough, block, and lot forms a unique key for property in New York City. Commonly called a BBL.
* `BUILDING CLASS AT PRESENT` and `BUILDING CLASS AT TIME OF SALE`: The type of building at various points in time.

Note that because this is a financial transaction dataset, there are some points that need to be kept in mind:

* Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
* This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

Formulate a question and derive a statistical hypothesis test to answer the question. You have to demonstrate that you’re able to make decisions using data in a scientific manner. Examples of questions can be:

* Is there a difference in unit sold between property built in 1900-2000 and 2001 so on?
* Is there a difference in unit sold based on building category?
* What can you discover about New York City real estate by looking at a year's worth of raw transaction records? Can you spot trends in the market?

Please make sure that you have completed the lesson for this course, namely Python and Practical Statistics which is part of this Program.

## Data Preparation

Load the library.

NYC property data from [here](https://www.kaggle.com/new-york-city/nyc-property-sales)

In [48]:
import math
import statistics
import scipy.stats
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [49]:
# Get your import statement here
df_prop = pd.read_csv("dataset/nyc-rolling-sales.csv")

df_prop

Unnamed: 0  BOROUGH   NEIGHBORHOOD  \
0               4        1  ALPHABET CITY   
1               5        1  ALPHABET CITY   
2               6        1  ALPHABET CITY   
3               7        1  ALPHABET CITY   
4               8        1  ALPHABET CITY   
...           ...      ...            ...   
84543        8409        5        WOODROW   
84544        8410        5        WOODROW   
84545        8411        5        WOODROW   
84546        8412        5        WOODROW   
84547        8413        5        WOODROW   

                           BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
0      07 RENTALS - WALKUP APARTMENTS                                2A   
1      07 RENTALS - WALKUP APARTMENTS                                 2   
2      07 RENTALS - WALKUP APARTMENTS                                 2   
3      07 RENTALS - WALKUP APARTMENTS                                2B   
4      07 RENTALS - WALKUP APARTMENTS                                2A   
...                                            ...                  ...   
84543  02 TWO FAMILY DWELLINGS                                        1   
84544  02 TWO FAMILY DWELLINGS                                        1   
84545  02 TWO FAMILY DWELLINGS                                        1   
84546  22 STORE BUILDINGS                                             4   
84547  35 INDOOR PUBLIC AND CULTURAL FACILITIES                       4   

       BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT                 ADDRESS  \
0        392    6                                  C2            153 AVENUE B   
1        399   26                                  C7   234 EAST 4TH   STREET   
2        399   39                                  C7   197 EAST 3RD   STREET   
3        402   21                                  C4     154 EAST 7TH STREET   
4        404   55                                  C2  301 EAST 10TH   STREET   
...      ...  ...       ...                       ...                     ...   
84543   7349   34                                  B9           37 QUAIL LANE   
84544   7349   78                                  B9        32 PHEASANT LANE   
84545   7351   60                                  B2        49 PITNEY AVENUE   
84546   7100   28                                  K6   2730 ARTHUR KILL ROAD   
84547   7105  679                                  P9       155 CLAY PIT ROAD   

       ... RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
0      ...                 5                 0            5              1633   
1      ...                28                 3           31              4616   
2      ...                16                 1           17              2212   
3      ...                10                 0           10              2272   
4      ...                 6                 0            6              2369   
...    ...               ...               ...          ...               ...   
84543  ...                 2                 0            2              2400   
84544  ...                 2                 0            2              2498   
84545  ...                 2                 0            2              4000   
84546  ...                 0                 7            7            208033   
84547  ...                 0                 1            1             10796   

       GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
0                   6440       1900                         2   
1                  18690       1900                         2   
2                   7803       1900                         2   
3                   6794       1913                         2   
4                   4615       1900                         2   
...                  ...        ...                       ...   
84543               2575       1998                         1   
84544               2377       1998                         1   
84545               1496       192

In [50]:
df_prop.columns

Index(['Unnamed: 0', 'BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'EASE-MENT',
       'BUILDING CLASS AT PRESENT', 'ADDRESS', 'APARTMENT NUMBER', 'ZIP CODE',
       'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       'SALE PRICE', 'SALE DATE'],
      dtype='object')

Let's view the top 5 rows of the dataset using the `head()` function.

In [51]:
df_prop.head(5)

Unnamed: 0  BOROUGH   NEIGHBORHOOD  \
0           4        1  ALPHABET CITY   
1           5        1  ALPHABET CITY   
2           6        1  ALPHABET CITY   
3           7        1  ALPHABET CITY   
4           8        1  ALPHABET CITY   

                       BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  BLOCK  \
0  07 RENTALS - WALKUP APARTMENTS                                2A    392   
1  07 RENTALS - WALKUP APARTMENTS                                 2    399   
2  07 RENTALS - WALKUP APARTMENTS                                 2    399   
3  07 RENTALS - WALKUP APARTMENTS                                2B    402   
4  07 RENTALS - WALKUP APARTMENTS                                2A    404   

   LOT EASE-MENT BUILDING CLASS AT PRESENT                 ADDRESS  ...  \
0    6                                  C2            153 AVENUE B  ...   
1   26                                  C7   234 EAST 4TH   STREET  ...   
2   39                                  C7   197 EAST 3RD   STREET  ...   
3   21                                  C4     154 EAST 7TH STREET  ...   
4   55                                  C2  301 EAST 10TH   STREET  ...   

  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
0                 5                 0            5              1633   
1                28                 3           31              4616   
2                16                 1           17              2212   
3                10                 0           10              2272   
4                 6                 0            6              2369   

   GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
0               6440       1900                         2   
1              18690       1900                         2   
2               7803       1900                         2   
3               6794       1913                         2   
4               4615       1900                         2   

   BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
0                              C2     6625000  2017-07-19 00:00:00  
1                              C7         -    2016-12-14 00:00:00  
2                              C7         -    2016-12-09 00:00:00  
3                              C4     3936272  2016-09-23 00:00:00  
4                              C2     8000000  2016-11-17 00:00:00  

[5 rows x 22 columns]

We can also veiw the bottom 5 rows of the dataset using the `tail()` function.

In [52]:
df_prop.tail(6)

Unnamed: 0  BOROUGH NEIGHBORHOOD  \
84542        8408        5      WOODROW   
84543        8409        5      WOODROW   
84544        8410        5      WOODROW   
84545        8411        5      WOODROW   
84546        8412        5      WOODROW   
84547        8413        5      WOODROW   

                           BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
84542  02 TWO FAMILY DWELLINGS                                        1   
84543  02 TWO FAMILY DWELLINGS                                        1   
84544  02 TWO FAMILY DWELLINGS                                        1   
84545  02 TWO FAMILY DWELLINGS                                        1   
84546  22 STORE BUILDINGS                                             4   
84547  35 INDOOR PUBLIC AND CULTURAL FACILITIES                       4   

       BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT                ADDRESS  \
84542   7339   41                                  B9         41 SONIA COURT   
84543   7349   34                                  B9          37 QUAIL LANE   
84544   7349   78                                  B9       32 PHEASANT LANE   
84545   7351   60                                  B2       49 PITNEY AVENUE   
84546   7100   28                                  K6  2730 ARTHUR KILL ROAD   
84547   7105  679                                  P9      155 CLAY PIT ROAD   

       ... RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
84542  ...                 2                 0            2              3020   
84543  ...                 2                 0            2              2400   
84544  ...                 2                 0            2              2498   
84545  ...                 2                 0            2              4000   
84546  ...                 0                 7            7            208033   
84547  ...                 0                 1            1             10796   

       GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
84542               1800       1997                         1   
84543               2575       1998                         1   
84544               2377       1998                         1   
84545               1496       1925                         1   
84546              64117       2001                         4   
84547               2400       2006                         4   

       BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
84542                              B9         -    2016-12-01 00:00:00  
84543                              B9      450000  2016-11-28 00:00:00  
84544                              B9      550000  2017-04-21 00:00:00  
84545                              B2      460000  2017-07-05 00:00:00  
84546                              K6    11693337  2016-12-21 00:00:00  
84547                              P9       69300  2016-10-27 00:00:00  

[6 rows x 22 columns]

BOROUGH: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).

In [53]:
# the dimensions of the dataframe

df_prop.shape

(84548, 22)

In [54]:
# Drop 'Unnamed: 0' and 'EASE-MENT' features
df_prop = df_prop.drop(columns={"Unnamed: 0","EASE-MENT"})
df_prop

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
...        ...            ...                                          ...   
84543        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84544        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84545        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84546        5        WOODROW  22 STORE BUILDINGS                            
84547        5        WOODROW  35 INDOOR PUBLIC AND CULTURAL FACILITIES      

      TAX CLASS AT PRESENT  BLOCK  LOT BUILDING CLASS AT PRESENT  \
0                       2A    392    6                        C2   
1                        2    399   26                        C7   
2                        2    399   39                        C7   
3                       2B    402   21                        C4   
4                       2A    404   55                        C2   
...                    ...    ...  ...                       ...   
84543                    1   7349   34                        B9   
84544                    1   7349   78                        B9   
84545                    1   7351   60                        B2   
84546                    4   7100   28                        K6   
84547                    4   7105  679                        P9   

                      ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
0                153 AVENUE B                      10009                  5   
1       234 EAST 4TH   STREET                      10009                 28   
2       197 EAST 3RD   STREET                      10009                 16   
3         154 EAST 7TH STREET                      10009                 10   
4      301 EAST 10TH   STREET                      10009                  6   
...                       ...              ...       ...                ...   
84543           37 QUAIL LANE                      10309                  2   
84544        32 PHEASANT LANE                      10309                  2   
84545        49 PITNEY AVENUE                      10309                  2   
84546   2730 ARTHUR KILL ROAD                      10309                  0   
84547       155 CLAY PIT ROAD                      10309                  0   

       COMMERCIAL UNITS  TOTAL UNITS LAND SQUARE FEET GROSS SQUARE FEET  \
0                     0            5             1633              6440   
1                     3           31             4616             18690   
2                     1           17             2212              7803   
3                     0           10             2272              6794   
4                     0            6             2369              4615   
...                 ...          ...              ...               ...   
84543                 0            2             2400              2575   
84544                 0            2             2498              2377   
84545                 0            2             4000              1496   
84546                 7            7           208033             64117   
84547                 1            1            10796              2400   

       YEAR BUILT  TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
0            1900                          2                             C2   
1            1900                          2                             C7   
2            1900                          2                             C7   
3            1913                          2                             C4   
4       

In [55]:
# Dtype of each features in dataframe 

df_prop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   BOROUGH                         84548 non-null  int64 
 1   NEIGHBORHOOD                    84548 non-null  object
 2   BUILDING CLASS CATEGORY         84548 non-null  object
 3   TAX CLASS AT PRESENT            84548 non-null  object
 4   BLOCK                           84548 non-null  int64 
 5   LOT                             84548 non-null  int64 
 6   BUILDING CLASS AT PRESENT       84548 non-null  object
 7   ADDRESS                         84548 non-null  object
 8   APARTMENT NUMBER                84548 non-null  object
 9   ZIP CODE                        84548 non-null  int64 
 10  RESIDENTIAL UNITS               84548 non-null  int64 
 11  COMMERCIAL UNITS                84548 non-null  int64 
 12  TOTAL UNITS                     84548 non-null

It looks like empty records are not being treated as NA. We convert columns to their appropriate data types to obtain NAs.

In [56]:
# First, let's check which columns should be categorical
print('Column name')
for col in df_prop.columns:
    if df_prop[col].dtype=='object':
        print(col, df_prop[col].nunique())

Column name
NEIGHBORHOOD 254
BUILDING CLASS CATEGORY 47
TAX CLASS AT PRESENT 11
BUILDING CLASS AT PRESENT 167
ADDRESS 67563
APARTMENT NUMBER 3989
LAND SQUARE FEET 6062
GROSS SQUARE FEET 5691
BUILDING CLASS AT TIME OF SALE 166
SALE PRICE 10008
SALE DATE 364


In [57]:
# LAND SQUARE FEET,GROSS SQUARE FEET, SALE PRICE, BOROUGH should be numeric. 
# SALE DATE datetime format.
# categorical: NEIGHBORHOOD, BUILDING CLASS CATEGORY, TAX CLASS AT PRESENT, BUILDING CLASS AT PRESENT,
# BUILDING CLASS AT TIME OF SALE, TAX CLASS AT TIME OF SALE,BOROUGH 

numer = ['LAND SQUARE FEET','GROSS SQUARE FEET', 'SALE PRICE', 'BOROUGH']
for col in numer: # coerce for missing values
    df_prop[col] = pd.to_numeric(df_prop[col], errors='coerce')

categ = ['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'BUILDING CLASS AT PRESENT', 'BUILDING CLASS AT TIME OF SALE', 'TAX CLASS AT TIME OF SALE']
for col in categ:
    df_prop[col] = df_prop[col].astype('category')

df_prop['SALE DATE'] = pd.to_datetime(df_prop['SALE DATE'], errors='coerce')
df_prop

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
...        ...            ...                                          ...   
84543        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84544        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84545        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84546        5        WOODROW  22 STORE BUILDINGS                            
84547        5        WOODROW  35 INDOOR PUBLIC AND CULTURAL FACILITIES      

      TAX CLASS AT PRESENT  BLOCK  LOT BUILDING CLASS AT PRESENT  \
0                       2A    392    6                        C2   
1                        2    399   26                        C7   
2                        2    399   39                        C7   
3                       2B    402   21                        C4   
4                       2A    404   55                        C2   
...                    ...    ...  ...                       ...   
84543                    1   7349   34                        B9   
84544                    1   7349   78                        B9   
84545                    1   7351   60                        B2   
84546                    4   7100   28                        K6   
84547                    4   7105  679                        P9   

                      ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
0                153 AVENUE B                      10009                  5   
1       234 EAST 4TH   STREET                      10009                 28   
2       197 EAST 3RD   STREET                      10009                 16   
3         154 EAST 7TH STREET                      10009                 10   
4      301 EAST 10TH   STREET                      10009                  6   
...                       ...              ...       ...                ...   
84543           37 QUAIL LANE                      10309                  2   
84544        32 PHEASANT LANE                      10309                  2   
84545        49 PITNEY AVENUE                      10309                  2   
84546   2730 ARTHUR KILL ROAD                      10309                  0   
84547       155 CLAY PIT ROAD                      10309                  0   

       COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
0                     0            5            1633.0             6440.0   
1                     3           31            4616.0            18690.0   
2                     1           17            2212.0             7803.0   
3                     0           10            2272.0             6794.0   
4                     0            6            2369.0             4615.0   
...                 ...          ...               ...                ...   
84543                 0            2            2400.0             2575.0   
84544                 0            2            2498.0             2377.0   
84545                 0            2            4000.0             1496.0   
84546                 7            7          208033.0            64117.0   
84547                 1            1           10796.0             2400.0   

       YEAR BUILT TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
0            1900                         2                             C2   
1            1900                         2                             C7   
2            1900                         2                             C7   
3            1913                         2                        

In [58]:
missing = df_prop.isnull().sum()/len(df_prop)*100

print(pd.DataFrame([missing[missing>0],pd.Series(df_prop.isnull().sum()[df_prop.isnull().sum()>1000])], index=['percent missing','how many missing']))

                  LAND SQUARE FEET  GROSS SQUARE FEET   SALE PRICE
percent missing          31.049818          32.658372     17.22217
how many missing      26252.000000       27612.000000  14561.00000


In [59]:
print(df_prop[(df_prop['LAND SQUARE FEET'].isnull()) & (df_prop['GROSS SQUARE FEET'].notnull())].shape)
print(df_prop[(df_prop['LAND SQUARE FEET'].notnull()) & (df_prop['GROSS SQUARE FEET'].isnull())].shape)

(6, 20)
(1366, 20)


There are 1372 rows that can be filled in with their approximate values.

In [60]:
df_prop['LAND SQUARE FEET'] = df_prop['LAND SQUARE FEET'].mask((df_prop['LAND SQUARE FEET'].isnull()) & (df_prop['GROSS SQUARE FEET'].notnull()), df_prop['GROSS SQUARE FEET'])
df_prop['GROSS SQUARE FEET'] = df_prop['GROSS SQUARE FEET'].mask((df_prop['LAND SQUARE FEET'].notnull()) & (df_prop['GROSS SQUARE FEET'].isnull()), df_prop['LAND SQUARE FEET'])

In [61]:
#  Check for duplicates before

print(sum(df_prop.duplicated()))

df_prop[df_prop.duplicated(keep=False)].sort_values(['NEIGHBORHOOD', 'ADDRESS']).head(10)

# df.duplicated() automatically excludes duplicates, to keep duplicates in df we use keep=False

# in df.duplicated(df.columns) we can specify column names to look for duplicates only in those mentioned columns.

765


BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
76286        5       ANNADALE  02 TWO FAMILY DWELLINGS                       
76287        5       ANNADALE  02 TWO FAMILY DWELLINGS                       
76322        5       ANNADALE  05 TAX CLASS 1 VACANT LAND                    
76323        5       ANNADALE  05 TAX CLASS 1 VACANT LAND                    
76383        5  ARDEN HEIGHTS  01 ONE FAMILY DWELLINGS                       
76384        5  ARDEN HEIGHTS  01 ONE FAMILY DWELLINGS                       
76643        5       ARROCHAR  02 TWO FAMILY DWELLINGS                       
76644        5       ARROCHAR  02 TWO FAMILY DWELLINGS                       
50126        4        ASTORIA  03 THREE FAMILY DWELLINGS                     
50127        4        ASTORIA  03 THREE FAMILY DWELLINGS                     

      TAX CLASS AT PRESENT  BLOCK  LOT BUILDING CLASS AT PRESENT  \
76286                    1   6350    7                        B2   
76287                    1   6350    7                        B2   
76322                   1B   6459   28                        V0   
76323                   1B   6459   28                        V0   
76383                    1   5741   93                        A5   
76384                    1   5741   93                        A5   
76643                    1   3103   57                        B2   
76644                    1   3103   57                        B2   
50126                    1    856  139                        C0   
50127                    1    856  139                        C0   

                   ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
76286    106 BENNETT PLACE                      10312                  2   
76287    106 BENNETT PLACE                      10312                  2   
76322  N/A HYLAN BOULEVARD                          0                  0   
76323  N/A HYLAN BOULEVARD                          0                  0   
76383       266 ILYSSA WAY                      10312                  1   
76384       266 ILYSSA WAY                      10312                  1   
76643  129 MC CLEAN AVENUE                      10305                  2   
76644  129 MC CLEAN AVENUE                      10305                  2   
50126  22-18 27TH   STREET                      11105                  3   
50127  22-18 27TH   STREET                      11105                  3   

       COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
76286                 0            2            8000.0             4208.0   
76287                 0            2            8000.0             4208.0   
76322                 0            0            6667.0             6667.0   
76323                 0            0            6667.0             6667.0   
76383                 0            1             500.0             1354.0   
76384                 0            1             500.0             1354.0   
76643                 0            2            5000.0             2733.0   
76644                 0            2            5000.0             2733.0   
50126                 0            3            2000.0             1400.0   
50127                 0            3            2000.0             1400.0   

       YEAR BUILT TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
76286        1985                         1                             B2   
76287        1985                         1                             B2   
76322           0                         1                             V0   
76323           0                         1                             V0   
76383        1996                         1                             A5   
76384        1996                         1                             A5   
76643        1925                         1                             B2   
76644        1925                         1                             B2   
50126       

In [62]:
# The dataframe has 765 duplicated rows (exluding the original rows).
df_prop.drop_duplicates(inplace=True)

print(sum(df_prop.duplicated()))

0


In [63]:
df_prop.columns

Index(['BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'BUILDING CLASS AT PRESENT',
       'ADDRESS', 'APARTMENT NUMBER', 'ZIP CODE', 'RESIDENTIAL UNITS',
       'COMMERCIAL UNITS', 'TOTAL UNITS', 'LAND SQUARE FEET',
       'GROSS SQUARE FEET', 'YEAR BUILT', 'TAX CLASS AT TIME OF SALE',
       'BUILDING CLASS AT TIME OF SALE', 'SALE PRICE', 'SALE DATE'],
      dtype='object')

In [64]:
df_prop.groupby('BUILDING CLASS CATEGORY').sum().index

CategoricalIndex(['01 ONE FAMILY DWELLINGS                    ',
                  '02 TWO FAMILY DWELLINGS                    ',
                  '03 THREE FAMILY DWELLINGS                  ',
                  '04 TAX CLASS 1 CONDOS                      ',
                  '05 TAX CLASS 1 VACANT LAND                 ',
                  '06 TAX CLASS 1 - OTHER                     ',
                  '07 RENTALS - WALKUP APARTMENTS             ',
                  '08 RENTALS - ELEVATOR APARTMENTS           ',
                  '09 COOPS - WALKUP APARTMENTS               ',
                  '10 COOPS - ELEVATOR APARTMENTS             ',
                  '11 SPECIAL CONDO BILLING LOTS              ',
                  '11A CONDO-RENTALS                           ',
                  '12 CONDOS - WALKUP APARTMENTS              ',
                  '13 CONDOS - ELEVATOR APARTMENTS            ',
                  '14 RENTALS - 4-10 UNIT                     ',
                  '15 CO

## Exploratory data analysis

Now, let's get a simple descriptive statistics for `COMMERCIAL UNITS` features.

In [65]:
df_prop[df_prop['COMMERCIAL UNITS']==0].describe()

BOROUGH         BLOCK           LOT      ZIP CODE  \
count  78777.000000  78777.000000  78777.000000  78777.000000   
mean       3.004329   4273.781015    395.422420  10722.737068   
std        1.298594   3589.241940    671.604654   1318.493961   
min        1.000000      1.000000      1.000000      0.000000   
25%        2.000000   1330.000000     23.000000  10304.000000   
50%        3.000000   3340.000000     52.000000  11209.000000   
75%        4.000000   6361.000000   1003.000000  11357.000000   
max        5.000000  16322.000000   9106.000000  11694.000000   

       RESIDENTIAL UNITS  COMMERCIAL UNITS   TOTAL UNITS  LAND SQUARE FEET  \
count       78777.000000           78777.0  78777.000000      5.278000e+04   
mean            1.691737               0.0      1.724133      3.140140e+03   
std             9.838994               0.0      9.835016      2.929999e+04   
min             0.000000               0.0      0.000000      0.000000e+00   
25%             0.000000               0.0      1.000000      1.600000e+03   
50%             1.000000               0.0      1.000000      2.295000e+03   
75%             2.000000               0.0      2.000000      3.300000e+03   
max           889.000000               0.0    889.000000      4.252327e+06   

       GROSS SQUARE FEET    YEAR BUILT    SALE PRICE  
count       5.278000e+04  78777.000000  6.562900e+04  
mean        2.714612e+03   1781.065451  9.952969e+05  
std         2.791294e+04    551.024570  3.329268e+06  
min         0.000000e+00      0.000000  0.000000e+00  
25%         9.750000e+02   1920.000000  2.400000e+05  
50%         1.600000e+03   1940.000000  5.294900e+05  
75%         2.388000e+03   1967.000000  9.219560e+05  
max         4.252327e+06   2017.000000  3.450000e+08

Let us try to understand the columns. Above table shows descriptive statistics for the numeric columns.

- There are zipcodes with 0 value
- Can block/lot numbers go up to 16322?
- Most of the properties have 2 unit and maximum of 1844 units? The latter might mean some company purchased a building. This should be treated as an outlier.
- Other columns also have outliers which needs further investigation.
- Year column has a year with 0
- Most sales prices less than 10000 can be treated as gift or transfer fees.

Descriptive statistics for `RESIDENTIAL UNITS` features

In [67]:
df_prop['RESIDENTIAL UNITS'].describe()

count    83783.000000
mean         1.999976
std         16.569118
min          0.000000
25%          0.000000
50%          1.000000
75%          2.000000
max       1844.000000
Name: RESIDENTIAL UNITS, dtype: float64

My findings :

Count total value of `BOROUGH` features

In [68]:
df_prop.value_counts('BOROUGH')

BOROUGH
4    26548
3    23843
1    18102
5     8296
2     6994
dtype: int64

The mean for each `Borough`

In [74]:
print('The mean for each Borough','\n')

df_prop.groupby('BOROUGH').mean().drop(columns={"ZIP CODE","YEAR BUILT"})

The mean for each Borough 



BLOCK         LOT  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
BOROUGH                                                                 
1        1107.657938  749.190421           2.276931          0.280521   
2        4202.934372  298.625679           3.343580          0.160280   
3        4317.583484  339.525228           2.011618          0.145451   
4        6630.747589  230.493672           1.687283          0.190410   
5        3261.786162  198.775554           1.230111          0.080400   

         TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    SALE PRICE  
BOROUGH                                                                  
1           2.597227       5646.945628       32622.997758  3.344642e+06  
2           3.510152       3909.012725        4489.147412  5.946771e+05  
3           2.194229       2016.057333        2481.756365  8.410834e+05  
4           1.908920       4265.245190        3345.841693  7.388827e+05  
5           1.313163       8339.843259        4709.373037  5.436113e+05

The `median` for each Borough

In [78]:
print('The median for each Borough','\n')
df_prop.groupby('BOROUGH').median().drop(columns={"ZIP CODE"})

The median for each Borough 



BLOCK     LOT  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
BOROUGH                                                                     
1        1170.0  1004.0                0.0               0.0          1.0   
2        4177.5    47.0                1.0               0.0          1.0   
3        4558.0    50.0                1.0               0.0          1.0   
4        5900.0    40.0                1.0               0.0          1.0   
5        3024.0    56.0                1.0               0.0          1.0   

         LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  SALE PRICE  
BOROUGH                                                               
1                  2498.0             7520.0      1937.0   1155000.0  
2                  2375.0             1764.0      1932.0    274897.0  
3                  1800.0             1520.0      1930.0    425605.0  
4                  2768.0             1728.0      1945.0    495000.0  
5                  3700.0             1632.0      1973.0    465000.0

The `mode` for each Borough

In [88]:
print('The median for each Borough','\n')
df_prop.groupby('BOROUGH').agg(pd.Series.mode)

The median for each Borough 



NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
BOROUGH                                                                         
1        UPPER EAST SIDE (59-79)  13 CONDOS - ELEVATOR APARTMENTS               
2                      RIVERDALE  02 TWO FAMILY DWELLINGS                       
3             BEDFORD STUYVESANT  02 TWO FAMILY DWELLINGS                       
4                 FLUSHING-NORTH  01 ONE FAMILY DWELLINGS                       
5                    GREAT KILLS  01 ONE FAMILY DWELLINGS                       

        TAX CLASS AT PRESENT  BLOCK  LOT BUILDING CLASS AT PRESENT  \
BOROUGH                                                              
1                          2     16    1                        R4   
2                          1   3944    1                        D4   
3                          1   2135    1                        R4   
4                          1   5066    1                        A1   
5                          1   2450    1                        A5   

                        ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
BOROUGH                                                                         
1        169 WEST 95TH   STREET                      10011                  0   
2           2287 JOHNSON AVENUE                      10462                  1   
3               429 KENT AVENUE                      11201                  1   
4            131-05 40TH   ROAD                      11354                  1   
5         1235 FOREST HILL ROAD                      10314                  1   

         COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
BOROUGH                                                                       
1                       0            1            2523.0           112850.0   
2                       0            1               0.0                0.0   
3                       0            1               0.0                0.0   
4                       0            1            4000.0             2000.0   
5                       0            1            4000.0             1200.0   

         YEAR BUILT  TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
BOROUGH                                                                         
1                 0                          2                             R4   
2                 0                          1                             D4   
3                 0                          1                             R4   
4              1920                          1                             A1   
5              1970                          1                             A5   

         SALE PRICE  SALE DATE  
BOROUGH                         
1              10.0 2017-08-07  
2               0.0 2017-06-29  
3               0.0 2016-10-06  
4              10.0 2016-11-10  
5              10.0 2017-03-01

Calculate the `Range` for each Borough

In [111]:
print('The range for each Borough','\n')
df_prop.groupby('BOROUGH')["SALE PRICE"].mode()

The range for each Borough 



AttributeError: 'SeriesGroupBy' object has no attribute 'mode'

Calculate the `SD` for each Borough

In [104]:
# Std yang dihitung merupakan std sample

print('The Standard Deviation for each Borough','\n')
df_prop.groupby('BOROUGH').std().drop(columns={"ZIP CODE", "YEAR BUILT"})

The Standard Deviation for each Borough 



BLOCK         LOT  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
BOROUGH                                                                 
1         537.249675  881.300458          22.279292          6.290587   
2        1056.766385  757.726557          22.331234          0.878369   
3        2708.396811  534.208901           8.541588          1.644420   
4        4327.260807  493.581339          18.127911         14.187361   
5        2318.962697  427.259707           2.308245          1.993219   

         TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    SALE PRICE  
BOROUGH                                                                  
1          23.283374      16514.096636      114888.218288  2.414048e+07  
2          22.681594      18371.742462       29467.900592  2.793509e+06  
3           8.707408       8878.556905       14867.597573  3.951912e+06  
4          23.147488      45746.045898       41695.613170  3.227420e+06  
5           3.027469      88303.260793       68926.020904  1.905544e+06

The `Variance` for each Borough.

In [105]:
# Variansi yang dihitung merupakan variansi sample

print('The Standard Deviation for each Borough','\n')
df_prop.groupby('BOROUGH').var(ddof=1).drop(columns={"ZIP CODE", "YEAR BUILT"})

The Standard Deviation for each Borough 



BLOCK            LOT  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
BOROUGH                                                                     
1        2.886372e+05  776690.497576         496.366872         39.571488   
2        1.116755e+06  574149.535826         498.684024          0.771532   
3        7.335413e+06  285379.149550          72.958719          2.704118   
4        1.872519e+07  243622.537761         328.621156        201.281217   
5        5.377588e+06  182550.857153           5.327995          3.972920   

         TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    SALE PRICE  
BOROUGH                                                                  
1         542.115484      2.727154e+08       1.319930e+10  5.827628e+14  
2         514.454709      3.375209e+08       8.683572e+08  7.803693e+12  
3          75.818955      7.882877e+07       2.210455e+08  1.561761e+13  
4         535.806222      2.092701e+09       1.738524e+09  1.041624e+13  
5           9.165570      7.797466e+09       4.750796e+09  3.631100e+12